EVG-19205 Support multi line selection for log lines and introduce a share menu

khelif96 opened the pull request:
EVG-19205

### Description 
First off apologies for the large PR but this is a complex feature. 

Adds an option to support multi-line selection via shift+clicking a range of log lines. This will open up a share menu with a few options 
- [x] Support multi-line selection with shift+click
- [x] Open menu automatically on share 
- [x] Support copying the selected range to the users clipboard. (Note this ignores filtered out lines)
- [x] Copy a link to the user's clipboard
- [x] Apply search range to only selected line range
- [x] Save selected range to URL for easier sharing
- [x] Add button/ui to open and close menu after its been closed if the range is still selected

### Screenshots
https://github.com/evergreen-ci/parsley/assets/4605522/e81250c2-b15f-4005-b58b-f29d4f329332
### Testing 
Several Unit tests
E2E tests in progress but should not impede initial reviews.

sophstad commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,138 @@
+    <Menu
+      data-cy="sharing-menu"
+      open={open}
+      trigger={
```
Scrolling the page leaves the menu behind, it would be nice for it to be attached to the trigger (though again this may be more of an LG issue)

sophstad replied:
Did this end up too difficult to implement?

khelif96 replied:
It did. Since this uses a portal to render this element it required using a ref on the scrollable container and propagating it down to the component. But since the scrollable container is handled internally by react-virtuoso it wasn't as easy to expose.

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,138 @@
+      : selectedLines.endingLine - selectedLines.startingLine + 1;
+
+  return (
+    <Menu
```
I tried really hard to get this working but it is a limitation of the LG popover component. You can either manually control the open state or have the on-outside click-close pattern but not both.  I opened this [ticket](https://jira.mongodb.org/browse/LG-3648) with LG to allow for both but without it, we will have to settle on one or the other.

khelif96 replied:
nvm fixed this thanks for the call out!

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,138 @@
+      }
+    >
+      <MenuItem
+        onClick={handleCopySelectedLines}
```
Good idea!

khelif96 commented on src/context/MultiLineSelectContext/index.tsx:
```diff
@@ -0,0 +1,122 @@
+
+type MultiLineSelectContextState = {
+  handleSelectLine: (selectedLine: number, shiftClick: boolean) => void;
+  clearSelection: () => void;
```
I had some other plans for it but I removed the export since I went in another direction.

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,138 @@
+      >
+        Copy selected {pluralize("line", lineCount)}
+      </MenuItem>
+      <MenuItem
```
Good catch I had planned to but forgot

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/SharingMenu.test.tsx:
```diff
@@ -0,0 +1,106 @@
+  it("should render an open menu", () => {
+    const { Component: MenuComponent } = renderComponentWithHook(
+      useMultiLineSelectContext,
+      <SharingMenuComponent defaultOpen />
```
I wonder what I was thinking when I wrote that 🤦

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/SharingMenu.test.tsx:
```diff
@@ -0,0 +1,106 @@
+      <SharingMenuComponent defaultOpen />
+    );
+    const { Component } = RenderFakeToastContext(<MenuComponent />);
+    const { router } = renderWithRouterMatch(<Component />, { wrapper });
```
Extracted it to a render function.

sophstad commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,147 @@
+            lineCount
+          )}.`}
+        >
+          Share link to selected {pluralize("line", lineCount)}
```
I think it may be helpful if it's clear that this is another copy action (as opposed to opening a new share menu, for example)
```suggestion
          Copy share link to selected {pluralize("line", lineCount)}
```

sophstad commented on src/context/MultiLineSelectContext/index.tsx:
```diff
@@ -0,0 +1,124 @@
+    initialMenuPosition ?? undefined
+  );
+
+  const clearSelection = useCallback(() => {
```
Right now (afaict) the only way to clear the selection is to click a few times: the first click closes the menu if it's open, the next replaces the selection range with a single selected line, and the third click clears that line.

Do you think a "clear selection" menu item would be helpful, or is that overkill? Or maybe clicking within the selection range should just clear the selection instead of replacing it with a single line selection?

khelif96 replied:
Just added a Clear selection menu item.

sophstad commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,147 @@
+        onClick={handleOnlySearchOnRange}
+        title="Limit the range Parsley will search to only these lines."
+      >
+        Only search on range
```
This is super extra, but it would be cool if clicking this button caused some kind of pulse or animation on the "Details" button to indicate that the range setting is there. If you like the idea could be worth just opening a ticket for.

khelif96 replied:
I think this is actually a great idea! I opened a follow-up ticket for it to do it this sprint since this pr is already pretty large. https://jira.mongodb.org/browse/DEVPROD-1647

sophstad commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,147 @@
+          lineCount
+        )} to your clipboard with JIRA formatting.`}
+      >
+        Copy selected {pluralize("line", lineCount)}
```
[nit] Do you think "Copy selected contents" would be clearer in terms of distinguishing it from the next option?

khelif96 replied:
Sure

khelif96 commented on cypress/integration/ansiLogs/ansi_logView.ts:
```diff
@@ -200,3 +200,64 @@ describe("expanding collapsed rows", () => {
+    cy.window().then((win) => {
+      win.navigator.clipboard.readText().then((text) => {
+        expect(text).to.eq(
+          `{noformat}\n[2022/03/02 17:01:58.587] Task logger initialized (agent version 2022-02-14 from 00a4c8f3e8e4559cc23e04a019b6d1725c40c3e5).\n...\n[2022/03/02 17:02:01.610] e391612 EVG-16049 Update spruce project page for admin only variables (#1114)\n[2022/03/02 17:02:01.610] 04a52b2 EVG-15959 Fix rerender method in test utils (#1118)\n...\n[2022/03/02 17:05:21.050] running setup group bec ...
```
`{noformat}` is part of the jira string formatting. https://github.com/evergreen-ci/parsley/blob/314a94f5881ee4695ae8cbed4ca9d25bb0a41c61/src/utils/string/index.ts#L23

khelif96 replied:
This is preexisting logic from the lobster days. Server engineers rely on this since copying log lines into Jira tickets is a common workflow. Without the formatting characters, the lines won't be rendered correctly in jira. 
<img width="345" alt="image" src="https://github.com/evergreen-ci/parsley/assets/4605522/13e20900-62fc-410b-9ae6-22758def76ef">

khelif96 replied:
We can open a ticket to investigate making this more generic in the future if it comes up.

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,154 @@
+      <MenuItem
+        disabled={lineCount === 1}
+        glyph={<Icon glyph="MagnifyingGlass" />}
+        onClick={handleOnlySearchOnRange}
```
Sophie also suggested the same thing above. I have a follow-up ticket to keep this pr smaller. https://jira.mongodb.org/browse/DEVPROD-1647 I will work on it immediately after merging this since this PR is already pretty complex.

khelif96 replied:
I don't consider "search on range" to be a filter instead it's more so the logic associated with how we perform searches in the app. We don't apply that logic to any of the other functionality located in the details menu so I don't necessarily thing search on range should be an exception. 
<img width="741" alt="image" src="https://github.com/evergreen-ci/parsley/assets/4605522/9ffe161c-3e7c-4b27-ba44-93d01d288bd9">

khelif96 replied:
If we do want to change this behavior we should apply the concept universally and it would probably be out of scope for this ticket since it would also affect how we render the navbar since its already pretty cramped.
<img width="1494" alt="image" src="https://github.com/evergreen-ci/parsley/assets/4605522/0a596f29-d903-4eef-bb55-ec79ff8b73e1">

khelif96 commented on package.json:
```diff
@@ -110,6 +111,7 @@
     "@types/lodash.debounce": "4.0.7",
     "@types/new-relic-browser": "0.1212.2",
     "@types/node": "16.11.47",
+    "@types/pluralize": "^0.0.31",
```
nope

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/index.tsx:
```diff
@@ -0,0 +1,154 @@
+        </MenuItem>
+      )}
+      <MenuItem
+        disabled={lineCount === 1}
```
That makes sense I enabled it

khelif96 commented:
> I discovered a bug where my clipboard is not overwritten when I try to copy the range for 1 line. You can reproduce this by doing just that. There is no success toast.

Fixed and added tests

khelif96 commented on src/components/LogRow/BaseRow/SharingMenu/utils.ts:
```diff
@@ -0,0 +1,30 @@
+  processedLogLines: ProcessedLogLines,
+  selectedLines: SelectedLineRange
+) => {
+  let { endingLine } = selectedLines;
+  if (endingLine === undefined) {
+    endingLine = selectedLines.startingLine;
+  }
```
Nice one!
